src/cmp.rs

```diff
@@ -1,38 +1,98 @@
+    #[derive(PartialEq)]
+    enum Origin {
+        Lhs,
+        Rhs,
+    }
+
+    // The loop below iterates through both iterators at once and handles ascii digits for comparison.
+    // If one iterator runs out of ascii digits, it is stored in this struct together with the
+    // information where it originated from.
+    struct NonDigit {
+        c: char,
+        origin: Origin,
+    }
+
+    impl core::ops::Deref for NonDigit {
+        type Target = char;
+
+        fn deref(&self) -> &Self::Target {
+            &self.c
+        }
+    }
+
+    impl NonDigit {
+        #[allow(dead_code)]
+
+        fn is_lhs(&self) -> bool {
+            self.origin == Origin::Lhs
+        }
+
+        fn is_rhs(&self) -> bool {
+            self.origin == Origin::Rhs
+        }
+    }
+
+    fn ok_if_ascii_digit(c: char) -> Result<char, char> {
+        Some(c).filter(char::is_ascii_digit).ok_or(c)
+    }
+
+    let mut current_cmp = None;
+    loop {
+        match (lhs.peek(), rhs.peek()) {
+            (Some(&a), Some(&b)) => {
+                let non_digit = match (ok_if_ascii_digit(a), ok_if_ascii_digit(b)) {
+                    (Ok(a), Ok(b)) => {
+                        // Only update current_cmp if the current comparison is yet undecided.
+                        // current_cmp is returned later when at least one iterator has hit a non-digit.
+                        if current_cmp.is_none() || current_cmp == Some(Ordering::Equal) {
+                            current_cmp = Some(a.cmp(&b));
+                        }
+                        None
+                    },
+                    (Err(c), Ok(_)) => Some(NonDigit{ c, origin: Origin::Lhs }),
+                    (Ok(_), Err(c)) => Some(NonDigit{ c, origin: Origin::Rhs }),
+                    (Err(_), Err(_)) => break current_cmp,
+                };
+
+                // Advance underlying iterators, since we only peek and break early if no iterator
+                // has any digits left, keeping these characters in the iterators for the caller to
+                // deal with in case current_cmp.is_none() or current_cmp == Some(Ordering::Equal).
+                let _ = lhs.next();
+                let _ = rhs.next();
+
+                // Return the appropriate ordering of a number versus non-digit characters.
+                if let Some(c) = non_digit {
+                    let mut ord = if current_cmp.is_none() && c.is_alphanumeric() {
+                        Ordering::Greater
                     } else {
-                        break;
+                        Ordering::Less
+                    };
+                    if c.is_rhs() {
+                        ord = ord.reverse();
                     }
+                    break Some(ord);
                 }
             }
+            (Some(_), None) => {
+                let _ = lhs.next();
```

Comment:
In practice, callers don't continue using the iterators when this returns an ordering != `Ordering::Equal`, so consuming the ascii digit in the non-exhausted iterator is not actually used for anything.
I just didn't feel comfortable to make this assumption in the function implementation, but I can remove it if you feel this is unnecessary.